### ISSUE TITLE
AST (Abstract Syntax Tree) needs to be linked to code for errors

### ISSUE THREAD

**Issue by lukasberglund** [ayazhafiz/hgb#58]
When we find type errors in the AST we need to be able to link them back to specific sections in the code. We have to find some way to make it clear which part of the code the specific parts of the AST correspond to.

**Comment by ayazhafiz** [comment]
A simple way is to attach parsed tokens, or at least their spans, to the grammar.

**Comment by lukasberglund** [comment]
Are you thinking add a token field to each expression constructor?

**Comment by ayazhafiz** [comment]
Yes, I think that is reasonable, or we could just pass the span of the token to the grammar.

For example, once we parse a string its span is actually different than the span of the string content we get while tokenizing.

```
"content"
^           Token.StrDelim
 ^^^^^^^    Token.String
        ^   Token.StrDelim
^^^^^^^^^   Grammar.String
```

**Comment by lukasberglund** [comment]
True!
